vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,469 @@
+	pool.mu.Lock()
+
+	go func() {
+		for ; ; time.Sleep(params.InitRetryDelay) {
+			err := initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			select {
+			case firstInitResult <- err:
+			default:
+			}
+			if err != nil {
+				continue
+			}
+			pool.mu.Unlock()
+
+			select {
+			case <-pool.close:
+				pool.proxy.terminate()
+				<-pool.proxy.onTerminated
+				pool.onClosed <- nil
+				return
+			case <-pool.proxy.onTerminated:
+			}
+
+			pool.mu.Lock()
+		}
+	}()
+
+	if err := <-firstInitResult; err != nil {
+		return nil, fmt.Errorf("can't initialize pool: %w", err)
+	}
+	return pool, nil
+}
+
+func (pool *pool) Count() int {
+	return pool.count
+}
+
+func (pool *pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	pool.mu.Lock()
+	proxy := pool.proxy
+	pool.mu.Unlock()
+
+	return proxy.CreateInstance(workdir, index)
+}
+
+// Close is not used now. Its support require wide code changes.
+// TODO: support the pool cleanup on syz-manager level.
+func (pool *pool) Close() error {
+	// TODO: pool.proxy.Call("Close")
+	pool.close <- true
+	<-pool.onClosed
+	return nil
+}
+
+type ProxyApp struct {
+	*rpc.Client
+	terminate    context.CancelFunc
+	onTerminated chan bool
+	subProcess   subProcessCmd // is needed for mocking
+	params       *proxyAppParams
+}
+
+type stdInOutCloser struct {
+	io.ReadCloser
+	io.WriteCloser
+}
+
+func (s stdInOutCloser) Close() error {
+	readErr := s.ReadCloser.Close()
+	writeErr := s.WriteCloser.Close()
+	if readErr != nil || writeErr != nil {
+		return fmt.Errorf("stdInOutCloser.Close() error on read: %v, write: %v", readErr, writeErr)
+	}
+	return nil
+}
+
+func runProxyApp(params *proxyAppParams, cmd string) (*ProxyApp, error) {
+	ctx, cancelContext := context.WithCancel(context.Background())
+	subProcess := params.CommandRunner(ctx, cmd)
+	toClose := []io.Closer{}
+	freeAll := func() {
+		for _, closer := range toClose {
+			closer.Close()
+		}
+		cancelContext()
+	}
+
+	subStdout, err := subProcess.StdoutPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdoutpipe: %w", err)
+	}
+	toClose = append(toClose, subStdout)
+
+	subStdin, err := subProcess.StdinPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdinpipe: %w", err)
+	}
+	toClose = append(toClose, subStdin)
+
+	subprocessLogs, err := subProcess.StderrPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stderrpipe: %w", err)
+	}
+	toClose = append(toClose, subprocessLogs)
+
+	codec := jsonrpc.NewClientCodec(stdInOutCloser{
+		subStdout,
+		subStdin,
+	})
+
+	go io.Copy(os.Stdout, subprocessLogs)
+
+	err = subProcess.Start()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to start command %v: %w", cmd, err)
+	}
+
+	onTerminated := make(chan bool, 1)
+
+	go func() {
+		err := subProcess.Wait()
+		if err != nil {
+			log.Logf(0, "failed to Wait() subprocess: %v", err)
+		}
+		freeAll()
```

Comment:
Docs say:
"Wait will close the pipe after seeing the command exit... It is thus incorrect to call Wait before all reads from the pipe have completed"
https://pkg.go.dev/os/exec#Cmd.StdoutPipe

We can get truncated stderr output (miss the actual error message when the subprocess crashes).

I think we need to call io.Copy(os.Stdout, subprocessLogs) in this goroutine before subProcess.Wait() and remove freeAll() call here.

For the same reason stdInOutCloser must not close r/w, they are already closed either by freeAll() or by subProcess.Wait().

You can do:
```
type stdInOutCloser struct {
	io.ReadCloser
	io.Writer
}

stdInOutCloser{io.NopCloser(subStdout), subStdin}
```

and remove the custom Close method.

Reply:
Good point.
But lets wait all read operations instead. And testing is not easy here...

Reply:
Waiting only subprocessLogs.
Additionally check that all the instances were closed.